lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,82 @@
+
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_filter( 'block_core_navigation_render_inner_blocks', 'gutenberg_hook_blocks_into_core_navigation', 10, 3 );
+
+/**
+ * Hooks into the REST API response for the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @param WP_REST_Request  $request  Request object.
+ * @return WP_REST_Response $response The response object.
+ */
+function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $post, $request ) {
+	$content                               = serialize_blocks( gutenberg_hook_first_last_children( $response->data['content']['raw'] ) );
+	$response->data['content']['raw']      = $content;
+	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
+
+	return $response;
+}
+
+/**
+ * Hooks into the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_Block_List  $inner_blocks The inner blocks of the core/navigation block.
+ * @return WP_Block_List $inner_blocks with the first and last inner blocks added.
+ */
+function gutenberg_hook_blocks_into_core_navigation( $inner_blocks ) {
+	return new WP_Block_List( gutenberg_hook_first_last_children( $inner_blocks ) );
+}
+
+/**
+ * Returns the first and last inner blocks of the core/navigation block.
+ *
+ * @param string|WP_Block_List $inner_blocks The inner blocks of the core/navigation block, or post_content string.
+ * @return array parsed inner blocks.
+ */
+function gutenberg_get_parsed_blocks( $blocks ) {
+	if ( is_string( $blocks ) ) {
+		return parse_blocks( $blocks );
+	}
+
+	if ( $blocks instanceof WP_Block_List ) {
+		return array_map(
+			function ( $inner_block ) {
+				return $inner_block->parsed_block;
+			},
+			iterator_to_array( $blocks )
+		);
+	}
+
+	return array();
+}
+
+/**
+ * Adds the first and last inner blocks if there are hooked blocks.
+ *
+ * @param string|WP_Block_List $inner_blocks The inner blocks of the core/navigation block, or post_content string.
+ * @return array $inner_blocks with the first and last inner blocks added.
+ */
+function gutenberg_hook_first_last_children( $inner_blocks ) {
+	$before_block_visitor = null;
+	$after_block_visitor  = null;
+	$hooked_blocks        = get_hooked_blocks();
+	$blocks               = gutenberg_get_parsed_blocks( $inner_blocks );
+
+	if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
+		$mock_anchor_parent_block = array(
+			'blockName'    => 'core/navigation',
+			'attrs'        => array(),
+			'innerBlocks'  => $blocks,
+			'innerContent' => array_fill( 0, count( $blocks ), null ),
+		);
+		$before_block_visitor     = make_before_block_visitor( $hooked_blocks, $mock_anchor_parent_block );
+		$after_block_visitor      = make_after_block_visitor( $hooked_blocks, $mock_anchor_parent_block );
```

Comment:
The second argument to these factories is `$context`, which isn't really used by any code, but only passed on to the [`hooked_block_types` filter](https://developer.wordpress.org/reference/hooks/hooked_block_types/) (so you could even omit it, and your code will still work).

I think it'd make sense to use it to pass the `wp_navigation` `WP_Post` object here, in analogy to how it's otherwise used:

> $context
> [WP_Block_Template](https://developer.wordpress.org/reference/classes/wp_block_template/)|array
> The block template, template part, or pattern that the anchor block belongs to.

Reply:
Just a couple of thoughts here.

I get argument errors when I remove this second argument: `Fatal error: Uncaught ArgumentCountError: Too few arguments to function make_before_block_visitor()`.

As for the `$context` should we not pass the anchor block as `$context` here so users can check it before adding their hooked block?

The problem with passing the `wp_navigation` `WP_Post` object here is that it won't have any of the `core/navigation` block attributes. [I have made a change to get the navigation block on the server-side](https://github.com/WordPress/gutenberg/pull/57754/commits/aeaa43a4cfd00f06fb23b7bc452243483dd85f82) if its an acceptable change, but still need to find a way to achieve this on the editor side.

What do you think to passing the entire Navigation block as context if we can make it possible?

Reply:
> Just a couple of thoughts here.
> 
> I get argument errors when I remove this second argument: `Fatal error: Uncaught ArgumentCountError: Too few arguments to function make_before_block_visitor()`.

Ah, my bad 😬 Yeah, I guess I couldn't just make them optional for the callbacks.

> As for the `$context` should we not pass the anchor block as `$context` here so users can check it before adding their hooked block?
>
> The problem with passing the `wp_navigation` `WP_Post` object here is that it won't have any of the `core/navigation` block attributes. [I have made a change to get the navigation block on the server-side](https://github.com/WordPress/gutenberg/pull/57754/commits/aeaa43a4cfd00f06fb23b7bc452243483dd85f82) if its an acceptable change, but still need to find a way to achieve this on the editor side.
> 
> What do you think to passing the entire Navigation block (even if its our mocked one) as context if we can make it possible?

Hmm, I'd really rather avoid passing a block as `$context`, TBH. I think it would muddle the concept quite a bit and be possibly confusing to people. Currently, while `$context` is a bit abstract, it's at least clear that it's _never_ a block -- we'd kinda lose that, and would end up with a situation where we might leave people guessing which argument to use to pass a block 😅

(I'm fairly confident that we don't _need_ it to pass the information we need; I'll elaborate on that in a separate comment.)